infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.devfile;
+
+import static java.lang.String.format;
+import static org.eclipse.che.api.workspace.server.devfile.Components.getIdentifiableComponentName;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.devfile.Component;
+import org.eclipse.che.api.core.model.workspace.devfile.Entrypoint;
+import org.eclipse.che.api.workspace.server.devfile.FileContentProvider;
+import org.eclipse.che.api.workspace.server.devfile.exception.DevfileException;
+import org.eclipse.che.api.workspace.server.devfile.exception.DevfileFormatException;
+import org.eclipse.che.api.workspace.server.devfile.validator.ComponentIntegrityValidator;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesRecipeParser;
+
+public class KubernetesComponentValidator implements ComponentIntegrityValidator {
+  private final KubernetesRecipeParser kubernetesRecipeParser;
+  private final Set<String> k8sBasedComponentTypes;
+
+  @Inject
+  public KubernetesComponentValidator(
+      KubernetesRecipeParser kubernetesRecipeParser,
+      @Named(KubernetesDevfileBindings.KUBERNETES_BASED_COMPONENTS_KEY_NAME)
+          Set<String> k8sBasedComponentTypes) {
+    this.kubernetesRecipeParser = kubernetesRecipeParser;
+    this.k8sBasedComponentTypes = k8sBasedComponentTypes;
+  }
+
+  @Override
+  public void validateComponent(Component component, FileContentProvider contentProvider)
+      throws DevfileFormatException {
+    try {
+      List<HasMetadata> selectedObjects = validateSelector(component, contentProvider);
+      validateEntrypointSelector(component, selectedObjects);
+    } catch (Exception e) {
+      throw new DevfileFormatException(
+          format(
+              "Failed to validate content reference of component '%s' of type '%s': %s",
+              getIdentifiableComponentName(component), component.getType(), e.getMessage()),
+          e);
+    }
+  }
+
+  /**
+   * Validates that the selector, if any, selects some objects from the component's referenced
```

Comment:
`component's referenced content` -> `components referenced content`?

Reply:
component's referenced content ~ referenced content of the component, so it is fine as is imho

Reply:
sorry, you're right.